src/core/packages/notifications/browser-internal/src/notification_coordinator.ts

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import * as Rx from 'rxjs';
+import { i18n } from '@kbn/i18n';
+
+export class Coordinator {
+  private readonly coordinationLock$ = new Rx.BehaviorSubject<{
+    locked: boolean;
+    controller: string | null;
+  }>({
+    locked: false,
+    controller: null,
+  });
+
+  public lock$ = this.coordinationLock$.asObservable();
+
+  public acquireLock(owner: string) {
+    const { locked, controller } = this.coordinationLock$.getValue();
+
+    if (locked && controller !== owner) {
+      throw new Error(
+        i18n.translate('core.notifications.notificationLockError', {
+          defaultMessage: 'Notification lock is already acquired',
+        })
+      );
+    }
+
+    this.coordinationLock$.next({ locked: true, controller: owner });
+  }
+
+  public releaseLock(owner: string) {
+    const { locked, controller } = this.coordinationLock$.getValue();
+
+    if (locked && controller === owner) {
+      this.coordinationLock$.next({ locked: false, controller: null });
+    }
+  }
+
+  /**
+   * @description This method is used to control the flow of notifications, it will automatically acquire
+   * a lock for the registered entity and buffer new values to be emitted when the provided condition is not met.
+   * Also when the passed observable has been emptied if the current lock still belongs to the registrar it will be released automatically, however an acquired lock can
+   * also be released as deemed fit using {@link releaseLock}.
+   * @param registrar - A unique identifier for the caller of this method
+   * @param $ - Observable to be controlled
+   * @param cond - Condition under which updates from the provided observable should be emitted
+   */
+  public optInToCoordination<T extends Array<{ id: string }>>(
+    registrar: string,
+    $: Rx.Observable<T>,
+    cond: Parameters<typeof Rx.filter<ReturnType<Coordinator['coordinationLock$']['getValue']>>>[0]
```

Comment:
I wonder if this can be simplified.

IMO, we can reuse the interface `NotificationCoordinatorState`

Which can be reused in line 14 when defining the `BehaviorSubject`. Then, `cond` can be defined as

```suggestion
    cond: (coordinationLock: NotificationCoordinatorState) => boolean
```